pdg/src/info.rs

```diff
@@ -0,0 +1,229 @@
+use crate::graph::{Graph, GraphId, Graphs, Node, NodeId, NodeKind};
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::Field;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt::{self, Debug, Display, Formatter};
+
+///The information checked in this struct is whether nodes flow to loads, stores, and offsets (pos
+///and neg), as well as whether they are unique.
+///Uniqueness of node X is determined based on whether there is a node Y which is X's ancestor
+///through copies, fields, and offsets, also is a node Z's ancestor through copies, fields, offsets,
+///where the fields are the same between X and Z, and where Z is chronologically between X and X's last descendent.
+///If such a node exists, X is not unique.
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct NodeInfo {
+    flows_to_mutation: Option<NodeId>,
+    flows_to_load: Option<NodeId>,
+    flows_to_pos_offset: Option<NodeId>,
+    flows_to_neg_offset: Option<NodeId>,
+    non_unique: Option<NodeId>,
+}
+
+impl Display for NodeInfo {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let fm = match self.flows_to_mutation {
+            None => "no mut".to_string(),
+            Some(x) => format!("mut {}", x),
+        };
+        let fl = match self.flows_to_load {
+            None => "no load".to_string(),
+            Some(x) => format!("load {}", x),
+        };
+        let fpo = match self.flows_to_pos_offset {
+            None => "no flow to pos_offset".to_string(),
+            Some(_) => "does flow to pos_offset".to_string(),
+        };
+        let fno = match self.flows_to_neg_offset {
+            None => "no flow to neg_offset".to_string(),
+            Some(_) => "does flow to neg_offset".to_string(),
+        };
+        let funi = match self.non_unique {
+            None => "unique".to_string(),
+            Some(x) => format!("non unique proven by {}", x),
+        };
+        write!(f, "{fm},{fl},{fpo},{fno},{funi}")
```

Comment:
This generates too much noise I think.  I think it's better if we only print the rarer cases (i.e. the `Some` cases).

Reply:
```suggestion
let s = [
            ("mut", self.flows_to_mutation),
            ("load", self.flows_to_load),
            ("+offset", self.flows_to_neg_offset),
            ("-offset", self.flows_to_neg_offset),
            ("non unique by", self.non_unique),
        ]
        .into_iter()
        .filter_map(|(name, node)| Some((name, node?)))
        .format_with(", ", |(name, node), f| f(&format_args!("{name} {node}")));
        write!(f, "{}", s)
```

How about this?

Reply:
Fix the snapshot before accepting this change, as it'll change the snapshot again and I want to see the diff.

Reply:
Sounds good. Do you think it should be abbreviated?